Including all SFA Functions

situx opened the pull request:
Beginning to add more SFA function definitions for GeoSPARQL 1.2
Closes #160
Closes #11

situx commented:
As far as the Simple Features functions are concerned, all functions have been added to this pull request. What remains to be done is to discuss the issues I pointed out in the first post, to resolve them and to review the work done.

mperry455 commented on 1.1/funcsrules.ttl:
```diff
@@ -662,14 +662,35 @@ funcs:asKML
+    fno:returns (funcs:asWKB_output ) ;
+    rdfs:isDefinedBy <http://www.opengis.net/spec/geosparql/1.2/req/geometry-extension/geometry-as-wkb-function> ;
+    rdfs:seeAlso <http://www.opengis.net/doc/geosparql/1.2> ;
+    skos:definition "Converts a given geometry to an equivalent WKT representation preserving the coordinate reference system and geometry type, if possible."@en ;
```
typo: "equivalent WKT representation" --> "equivalent WKB representation"

situx replied:
Fixed

mperry455 commented on 1.1/geo.ttl:
```diff
@@ -794,6 +806,18 @@ PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
 		spec11:B.1.2.3 ;		
 .
 
+:isEmpty
```
Do you mean isClosed?

situx replied:
Yes. It is fixed now

mperry455 commented on 1.1/spec/11-Part-08.adoc:
```diff
@@ -1157,7 +1388,24 @@ The function http://www.opengis.net/def/function/geosparql/perimeter[`geof:perim
+The function http://www.opengis.net/def/function/geosparql/pointN[`geof:pointN`] Returns the nth point in `geom`.
+
+
+==== Function: geof:pointOnSurface
```
Do we need this one in funcsrules.ttl?

mperry455 commented on 1.1/spec/11-Part-08.adoc:
````diff
@@ -904,6 +1054,14 @@ geof:boundingCircle (geom: ogc:geomLiteral): ogc:geomLiteral
+geof:boundingPolygons (geom: ogc:geomLiteral, n: xsd:integer): ogc:geomLiteral
+```
+
+The function http://www.opengis.net/def/function/geosparql/boundingPolygons[`geof:boundingPolygons`] returns a collection of polygons which bound the nth polgon of a given Multipolygon `geom`.
````
Multipolygon --> PolyhedralSurface?

situx replied:
Yes, that fits better. Fixed now

situx commented:
There is another issue we need to consider here:
https://libgeos.org/specifications/wkb/
As you can see, WKB is not sufficient to encode 3D geometries.
To allow for the encoding of 3D geometries we need to allow EWKB (Extended WKB), so that is what I will propose in this pull request instead.

nicholascar commented:
I don't have experience with the use of WKB, TinyWKB or EWKB, or experience with the community that uses them, to know of the pros and cons here.

My quick reading of the format descriptors seems to indicate TinyWKB is a lot better than WKB but that's just from reading the spec, not use.

Since we don't have a legacy of supporting any WKB format so far, and we are very interested in 3D support in this standard version, I suggest we go for the formats that support 3D, so this is TinyWKB & EWKB.

For WKT, we actually already support a form of EWTK now, since we specify how to include the SRID.

Perhaps we should clarify a policy on literal definitions: do we try and define as many as possible, like 3 x WKB, 2 x WKT, and leave it up to implementers to actually realise or do we try to pick some winners like perhaps only EWKB & EWKT, stealing from Postgres?

situx commented:
In today's meeting we thought that the definition of EKWB would be enough since it covers 3D and all WKB definitions. TWKB did not seem to be widely used.
I will adjust the specification accordingly so we only define one EWKB literal with Base64 encoding with a URI definition preceding the Base64 encoded EWKB String.
With that the pull request would again be open for review and approvals/rejects. 
It will be merged once we have established in which version/branch of the document to merge it.

situx commented:
I updated the pull request yet again and I think it is final now.
Changes:

- I only defined one wkbLiteral now and allow for it to use the types which are defined in ExtendedWKB (the types that define 3D geometries)
- I recommend not adding an SRID into the literal. WKB does not define it, EKWB technically allows it, but I wrote in the standard that it will be ignored. The SRID should be defined as a URI like in WKT or - in the future - maybe in the knowledge graph itself
- I fixed the comments of Matt
With that this pull request is again ready for review and once approved we can merge it in wherever we maintain the sepc for 1.2 onwards

jabhay commented:
We discussed the label for WKB and decided it would be better to changed it to EWKB to ensure that people were aware that it included EWKB and WKB as a proper subset of EWKB.

- [ ] Change to ewkbLiteral

situx commented:
All done. geo:wkbLiteral is renamed to geo:ewkbLiteral
